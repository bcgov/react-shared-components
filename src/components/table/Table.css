.bcgov-table {
  width: 100%;
}

.bcgov-table-body {
  display: flex;
  flex-direction: column;
}

.bcgov-row {
  display: flex;
  flex-direction: row;
  vertical-align: top;
  padding-bottom: 3px;
  padding-top: 3px;
  padding-right: 3px;
  white-space: pre-wrap;
}

.bcgov-blue-stripe .bcgov-table-body .bcgov-row:nth-child(even) {
  background: rgba(241, 244, 250, 1);
}

.bcgov-blue-stripe .bcgov-table-body .bcgov-row:nth-child(odd) {
  background: rgba(225, 232, 242, 1);
}

.bcgov-side-by-side {
  width: 60%;
}

.bcgov-vertical-middle {
  width: 35%;
  vertical-align: middle !important;
}

.bcgov-reduce-width {
  width: 20%;
}

.bcgov-empty-row td {
  border-bottom: 1px solid black;
}

.bcgov-right-align {
  text-align: right;
}

@media (max-width: 576px) {
  .bcgov-row {
    flex-direction: column;
  }

  .bcgov-side-by-side,
  .bcgov-vertical-middle,
  .bcgov-reduce-width {
    width: 100%;
  }
}
