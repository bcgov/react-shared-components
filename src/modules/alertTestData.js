import React from "react";
import { Alert } from "../components/alert/Alert";

export function generateAlert(icon, type, styling, element) {
  return <Alert icon={icon} type={type} styling={styling} element={element} />;
}
