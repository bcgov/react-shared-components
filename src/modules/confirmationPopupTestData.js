/* eslint-disable react/jsx-one-expression-per-line */
import React from "react";

const body = () => (
  <>
    <p>Your files will not be submitted.</p>
    <p>
      You will be returned to:
      <br />
      <b>Original</b> website
    </p>
  </>
);

export function getConfirmationPopupTestData() {
  return { body };
}
