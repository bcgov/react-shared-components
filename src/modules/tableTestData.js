const elements = [
  { name: "BCeID", value: "bobross42" },
  {
    name: "Last Name",
    value: "Ross",
  },
  { name: "First Name", value: "Bob" },
  { name: "Email Address", value: "bob.ross@example.com" },
];

export function getTableElementsTestData() {
  return elements;
}
