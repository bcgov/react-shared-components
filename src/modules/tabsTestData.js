import React from "react";

export const tabs = [
  {
    title: "Tab1",
    content: <p> Tab1 Content... </p>,
  },
  {
    title: "Tab2",
    content: <p> Tab2 Content... </p>,
  },
  {
    title: "Tab3",
    content: <p> Tab3 Content... </p>,
    isDisabled: true,
  },
];
