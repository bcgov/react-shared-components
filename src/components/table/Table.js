import React from "react";
import PropTypes from "prop-types";

import "./Table.css";

const TableElement = ({
  element: { name, value, isValueBold, isNameBold, isEmptyRow },
  isFeesData,
  elementStyles,
}) => {
  const emptyRow = isEmptyRow ? "bcgov-empty-row" : "";
  const rightAlign = isFeesData ? "bcgov-right-align" : "";
  let rowStyle = isFeesData ? "bcgov-row-fees" : "bcgov-row";
  let columnStyle = "bcgov-fill-width";
  if (elementStyles) {
    if (elementStyles.rowStyle) rowStyle += ` ${elementStyles.rowStyle}`;
    if (elementStyles.columnStyle)
      columnStyle = elementStyles && elementStyles.columnStyle;
  }

  return (
    <div className={`${rowStyle} ${emptyRow}`}>
      {isNameBold && (
        <div className={columnStyle}>
          <b>{name}</b>
        </div>
      )}
      {!isNameBold && <div className={columnStyle}>{name}</div>}
      {isValueBold && (
        <div className={`bcgov-table-value ${rightAlign}`}>
          <b>{value}</b>
        </div>
      )}
      {!isValueBold && (
        <div className={`bcgov-table-value ${rightAlign}`}>{value}</div>
      )}
    </div>
  );
};

export const Table = ({
  heading,
  elements,
  styling,
  isFeesData,
  elementStyles,
}) => {
  const tableComponents = elements.map((element) => {
    return (
      <TableElement
        isFeesData={isFeesData}
        key={element.key || element.name}
        element={element}
        elementStyles={elementStyles}
      />
    );
  });

  return (
    <div className={`bcgov-table ${styling}`}>
      <b>{heading}</b>
      <div className={"bcgov-table-body"}>{tableComponents}</div>
    </div>
  );
};

TableElement.propTypes = {
  element: PropTypes.shape({
    name: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    isValueBold: PropTypes.bool,
    isNameBold: PropTypes.bool,
    isEmptyRow: PropTypes.bool,
  }).isRequired,
  isFeesData: PropTypes.bool.isRequired,
  elementStyles: PropTypes.object,
};

Table.propTypes = {
  heading: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  elements: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
        .isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
        .isRequired,
      isValueBold: PropTypes.bool,
      isNameBold: PropTypes.bool,
      isEmptyRow: PropTypes.bool,
    }).isRequired
  ),
  styling: PropTypes.string,
  isFeesData: PropTypes.bool,
  elementStyles: PropTypes.object,
};

Table.defaultProps = {
  styling: "",
  heading: "",
  elements: [],
  isFeesData: false,
  elementStyles: {},
};
