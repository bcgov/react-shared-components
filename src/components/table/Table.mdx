import { Meta, Story } from "@storybook/addon-docs/blocks";
import { Table } from "./Table";

<Meta title="Table" component={Table} />

# Table

The table lets you aggregate a huge amount of data and present it in a clear and orderly way.

## API

| Prop          | Type             | Required | Description                                             | Default |
| ------------- | ---------------- | -------- | ------------------------------------------------------- | ------- |
| heading       | string/object    | false    | value representing heading to display within table      | ""      |
| elements      | array of objects | false    | content of the table                                    | []      |
| styling       | string           | false    | represents the styling you wish to apply for your table | ""      |
| elementStyles | object           | false    | represents row and column styles you wish to apply      | {}      |

### Elements API

The elements array contains objects of the following structure:

| Prop        | Type          | Required | Description                                                  | Default |
| ----------- | ------------- | -------- | ------------------------------------------------------------ | ------- |
| name        | string/object | true     | the key of the key:value pair in the table                   |         |
| value       | string/object | true     | the value of the key:value pair in the table                 |         |
| isValueBold | boolean       | false    | represents if you want your values to be bolded in the table | false   |
| isNameBold  | boolean       | false    | represents if you want your keys to be bolded in the table   | false   |
| isEmptyRow  | boolean       | false    | represents if you want to have an empty row with hr tag      | false   |

<br />

## Without Header

A table without a header.

<Story name="Without Header" id="table--without-header"></Story>

```bash
const elements = [
  { name: "BCeID", value: "bobross42" },
  {
    name: "Last Name",
    value: "Ross"
  },
  { name: "First Name", value: "Bob" },
  { name: "Email Address", value: "bob.ross@example.com" }
];

<Table elements={elements} />;
```

## With Header

A table with a header.

<Story name="With Header" id="table--with-header"></Story>

```bash
<Table heading="BCeID Info" elements={elements} />;
```

## With Blue Stripe

A table with blue stripes.

<Story name="With Blue Stripe" id="table--with-blue-stripe"></Story>

```bash
<Table elements={elements} styling="blue-stripe" />
```

## With Bold Values

A table with bolded values.

<Story name="With Bold Values" id="table--with-bold-values"></Story>

```bash
const tableDataWithBoldValues = [
  { name: "Key", value: "Value", isValueBold: true }
];

<Table elements={tableDataWithBoldValues} />
```

## With Bold Names

A table with bolded names.

<Story name="With Bold Names" id="table--with-bold-names"></Story>

```bash
const tableDataWithBoldNames = [
  { name: "Key", value: "Value", isNameBold: true }
];

<Table elements={tableDataWithBoldNames} />
```

## With Fees

A table for displaying fees.

<Story name="With Fees" id="table--with-fees"></Story>

```bash
const feesData = [
  { name: "Some Fees:", value: "$100.00", isValueBold: true },
  { name: "Some More Fees:", value: "$10.00", isValueBold: true }
];

<Table elements={feesData} />
```
