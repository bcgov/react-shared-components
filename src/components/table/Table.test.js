import React from "react";
import testBasicSnapshot from "../../TestHelper";

import { Table } from "./Table";
import { getTableElementsTestData } from "../../modules/tableTestData";

const tableData = getTableElementsTestData();
const header = "BCeID Info";

describe("Table Component", () => {
  test("Matches the default snapshot", () => {
    testBasicSnapshot(<Table heading={header} elements={tableData} />);
  });

  test("Matches the blue striped snapshot", () => {
    testBasicSnapshot(
      <Table elements={tableData} styling="bcgov-blue-stripe" />
    );
  });
});
