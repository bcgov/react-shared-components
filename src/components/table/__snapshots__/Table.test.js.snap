// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component Matches the blue striped snapshot 1`] = `
<div
  className="bcgov-table bcgov-blue-stripe"
>
  <b>
    
  </b>
  <div
    className="bcgov-table-body"
  >
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        BCeID
      </div>
      <div>
        bobross42
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        Last Name
      </div>
      <div>
        Ross
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        First Name
      </div>
      <div>
        Bob
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        Email Address
      </div>
      <div>
        bob.ross@example.com
      </div>
    </div>
  </div>
</div>
`;

exports[`Table Component Matches the default snapshot 1`] = `
<div
  className="bcgov-table "
>
  <b>
    BCeID Info
  </b>
  <div
    className="bcgov-table-body"
  >
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        BCeID
      </div>
      <div>
        bobross42
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        Last Name
      </div>
      <div>
        Ross
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        First Name
      </div>
      <div>
        Bob
      </div>
    </div>
    <div
      className="bcgov-row "
    >
      <div
        className="bcgov-vertical-middle "
      >
        Email Address
      </div>
      <div>
        bob.ross@example.com
      </div>
    </div>
  </div>
</div>
`;
